src/allmydata/test/test_grid_manager.py

```diff
@@ -0,0 +1,446 @@
+        }
+
+        fp.makedirs()
+        with fp.child("config.json").open("w") as f:
+            f.write(json.dumps_bytes(config))
+        with fp.child("alice.cert.0").open("w") as f:
+            f.write(json.dumps_bytes(bad_cert))
+
+        with self.assertRaises(ValueError) as ctx:
+            load_grid_manager(fp)
+
+        self.assertIn(
+            "Unknown certificate version",
+            str(ctx.exception),
+        )
+
+    def test_invalid_no_private_key(self):
+        """
+        Invalid Grid Manager config with no private key
+        """
+        tempdir = self.mktemp()
+        fp = FilePath(tempdir)
+        bad_config = {
+            "grid_manager_config_version": 0,
+        }
+        fp.makedirs()
+        with fp.child("config.json").open("w") as f:
+            f.write(json.dumps_bytes(bad_config))
+
+        with self.assertRaises(ValueError) as ctx:
+            load_grid_manager(fp)
+        self.assertIn(
+            "'private_key' required",
+            str(ctx.exception),
+        )
+
+    def test_invalid_bad_private_key(self):
+        """
+        Invalid Grid Manager config with bad private-key
+        """
+        tempdir = self.mktemp()
+        fp = FilePath(tempdir)
+        bad_config = {
+            "grid_manager_config_version": 0,
+            "private_key": "not actually encoded key",
+        }
+        fp.makedirs()
+        with fp.child("config.json").open("w") as f:
+            f.write(json.dumps_bytes(bad_config))
+
+        with self.assertRaises(ValueError) as ctx:
+            load_grid_manager(fp)
+        self.assertIn(
+            "Invalid Grid Manager private_key",
+            str(ctx.exception),
+        )
+
+    def test_invalid_storage_server(self):
+        """
+        Invalid Grid Manager config with missing public-key for
+        storage-server
+        """
+        tempdir = self.mktemp()
+        fp = FilePath(tempdir)
+        bad_config = {
+            "grid_manager_config_version": 0,
+            "private_key": "priv-v0-ub7knkkmkptqbsax4tznymwzc4nk5lynskwjsiubmnhcpd7lvlqa",
+            "storage_servers": {
+                "bad": {}
+            }
+        }
+        fp.makedirs()
+        with fp.child("config.json").open("w") as f:
+            f.write(json.dumps_bytes(bad_config))
+
+        with self.assertRaises(ValueError) as ctx:
+            load_grid_manager(fp)
+        self.assertIn(
+            "No 'public_key' for storage server",
+            str(ctx.exception),
+        )
+
+    def test_parse_cert(self):
+        """
+        Parse an ostensibly valid storage certificate
+        """
+        js = parse_grid_manager_certificate('{"certificate": "", "signature": ""}')
+        self.assertEqual(
+            set(js.keys()),
+            {"certificate", "signature"}
+        )
+        # the signature isn't *valid*, but that's checked in a
+        # different function
```

Comment:
It seems like it could very easily be the "parse" function's job to make sure the certificate and signature values at least ... parse.  But as long as something does I guess it's not a very big deal right now.

Reply:
Weeellll, yeah. But then the "parse" has to be given the correct keys, and "do crypto" which arguably is beyond "syntactic validation"?

Reply:
I just mean it would be nice if it were structurally valid coming out of the parser.  "" is neither a valid certificate or signature and cannot ever be, regardless of which keys are involved.

So more concretely this would be something like: (a) the values are strings (b) the strings have the correct length and (c) the strings are composed of legal characters for the encoding (base32, looks like).

If any of these don't hold then _most likely_ the next piece of the implementation that gets its hands on them will blow up and the high-level result is the same - the certificate is not used, the server is not uploaded to.  A failure that rejects the bogus inputs at parse time as structurally invalid will probably produce a better user experience than some other error that just happens because some piece of the implementation tripped over a value of the wrong type / size / encoding / etc.

It's fine with me if this is addressed in a follow-up ticket.